src/node_modules.cc

```diff
@@ -0,0 +1,412 @@
+    }
+
+    return std::nullopt;
+  };
+
+  if (error || document.get_object().get(main_object)) {
+    return throw_invalid_package_config();
+  }
+
+  simdjson::ondemand::raw_json_string key;
+  simdjson::ondemand::value value;
+  std::string_view field_value;
+  simdjson::ondemand::json_type field_type;
+  PackageConfig package_config{};
+  package_config.file_path = path;
+  package_config.raw_json = std::move(input);
+
+  for (auto field : main_object) {
+    // Throw error if getting key or value fails.
+    if (field.key().get(key) || field.value().get(value)) {
+      return throw_invalid_package_config();
+    }
+
+    if (key == "name") {
+      // Though there is a key "name" with a corresponding value,
+      // the value may not be a string or could be an invalid JSON string
+      if (value.get_string(package_config.name)) {
+        return throw_invalid_package_config();
+      }
+    } else if (key == "main") {
+      if (value.get_string(package_config.main)) {
+        return throw_invalid_package_config();
+      }
+    } else if (key == "exports") {
+      if (value.type().get(field_type)) {
+        return throw_invalid_package_config();
+      }
+      switch (field_type) {
+        case simdjson::ondemand::json_type::object:
+        case simdjson::ondemand::json_type::array: {
+          if (value.raw_json().get(field_value)) {
+            return throw_invalid_package_config();
+          }
+          package_config.exports = std::string(field_value);
+          break;
+        }
+        case simdjson::ondemand::json_type::string: {
+          if (value.get_string(package_config.exports)) {
+            return throw_invalid_package_config();
+          }
+          break;
+        }
+        default:
+          break;
+      }
+    } else if (key == "imports") {
+      if (value.type().get(field_type)) {
+        return throw_invalid_package_config();
+      }
+      switch (field_type) {
+        case simdjson::ondemand::json_type::array:
+        case simdjson::ondemand::json_type::object: {
+          if (value.raw_json().get(field_value)) {
+            return throw_invalid_package_config();
+          }
+          package_config.imports = std::string(field_value);
+          break;
+        }
+        case simdjson::ondemand::json_type::string: {
+          if (value.get_string(package_config.imports)) {
+            return throw_invalid_package_config();
+          }
+          break;
+        }
+        default:
+          break;
+      }
+    } else if (key == "type") {
+      if (value.get_string().get(field_value)) {
+        return throw_invalid_package_config();
+      }
+      // Only update type if it is "commonjs" or "module"
+      // The default value is "none" for backward compatibility.
+      if (field_value == "commonjs" || field_value == "module") {
+        package_config.type = std::string(field_value);
+      }
+    }
+  }
+
+  binding_data->package_configs_.insert({path, package_config});
```

Comment:
If you return a pointer, you could move package_config in cache and gain some performance.

Reply:
https://stackoverflow.com/questions/14816409/moving-an-object-into-a-map